In [73]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [74]:
df = pd.read_csv('../dataframes/online_advertising_performance_data.csv')
df.head()

month  day campaign_number user_engagement     banner placement  displays  \
0  April    1          camp 1            High  160 x 600       abc         4   
1  April    1          camp 1            High  160 x 600       def     20170   
2  April    1          camp 1            High  160 x 600       ghi     14701   
3  April    1          camp 1            High  160 x 600       mno    171259   
4  April    1          camp 1             Low  160 x 600       def       552   

       cost  clicks   revenue  post_click_conversions  \
0    0.0060       0    0.0000                       0   
1   26.7824     158   28.9717                      23   
2   27.6304     158   28.9771                      78   
3  216.8750    1796  329.4518                     617   
4    0.0670       1    0.1834                       0   

   post_click_sales_amount  Unnamed: 12  Unnamed: 13  
0                   0.0000          NaN          NaN  
1                1972.4602          NaN          NaN  
2                2497.2636          NaN          NaN  
3               24625.3234          NaN          NaN  
4                   0.0000          NaN          NaN

In [75]:
df.shape

(15408, 14)

**Storing the copy of the dataset.**

In [76]:
df_copy = df.copy()

In [77]:
## Exporting the dataset for manual analysis into the excel
with pd.ExcelWriter('digital_marketing.xlsx') as writer:
    df.to_excel(writer,sheet_name='page1')

**Data Cleaning**

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15408 entries, 0 to 15407
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   month                    15408 non-null  object 
 1   day                      15408 non-null  int64  
 2   campaign_number          15408 non-null  object 
 3   user_engagement          15408 non-null  object 
 4   banner                   15408 non-null  object 
 5   placement                14995 non-null  object 
 6   displays                 15408 non-null  int64  
 7   cost                     15408 non-null  float64
 8   clicks                   15408 non-null  int64  
 9   revenue                  15408 non-null  float64
 10  post_click_conversions   15408 non-null  int64  
 11  post_click_sales_amount  15408 non-null  float64
 12  Unnamed: 12              0 non-null      float64
 13  Unnamed: 13              0 non-null      float64
dtypes: float64(5), int64(4

**The Unnamed: 12 and Unnamed: 13 has zero non-null values. It would be good to remove them from the dataset to clean the dataset.**

In [79]:
df.drop(df.iloc[:, 12:13], inplace=True, axis=1)

In [80]:
df.drop(df.iloc[:, 12:13], inplace=True, axis=1)


In [81]:
df.head()

month  day campaign_number user_engagement     banner placement  displays  \
0  April    1          camp 1            High  160 x 600       abc         4   
1  April    1          camp 1            High  160 x 600       def     20170   
2  April    1          camp 1            High  160 x 600       ghi     14701   
3  April    1          camp 1            High  160 x 600       mno    171259   
4  April    1          camp 1             Low  160 x 600       def       552   

       cost  clicks   revenue  post_click_conversions  post_click_sales_amount  
0    0.0060       0    0.0000                       0                   0.0000  
1   26.7824     158   28.9717                      23                1972.4602  
2   27.6304     158   28.9771                      78                2497.2636  
3  216.8750    1796  329.4518                     617               24625.3234  
4    0.0670       1    0.1834                       0                   0.0000

In [82]:
df.shape

(15408, 12)

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15408 entries, 0 to 15407
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   month                    15408 non-null  object 
 1   day                      15408 non-null  int64  
 2   campaign_number          15408 non-null  object 
 3   user_engagement          15408 non-null  object 
 4   banner                   15408 non-null  object 
 5   placement                14995 non-null  object 
 6   displays                 15408 non-null  int64  
 7   cost                     15408 non-null  float64
 8   clicks                   15408 non-null  int64  
 9   revenue                  15408 non-null  float64
 10  post_click_conversions   15408 non-null  int64  
 11  post_click_sales_amount  15408 non-null  float64
dtypes: float64(3), int64(4), object(5)
memory usage: 1.4+ MB


In [84]:
df[df['placement'].isnull()]

month  day campaign_number user_engagement     banner placement  \
893    April    5          camp 1            High   728 x 90       NaN   
951    April    5          camp 2             Low   728 x 90       NaN   
957    April    5          camp 2          Medium   728 x 90       NaN   
1034   April    5          camp 3          Medium   728 x 90       NaN   
1042   April    6          camp 1            High  160 x 600       NaN   
...      ...  ...             ...             ...        ...       ...   
15259   June   29          camp 3             Low   728 x 90       NaN   
15265   June   29          camp 3          Medium   728 x 90       NaN   
15323   June   30          camp 1            High   728 x 90       NaN   
15391   June   30          camp 3            High   728 x 90       NaN   
15402   June   30          camp 3          Medium   728 x 90       NaN   

       displays    cost  clicks  revenue  post_click_conversions  \
893           8  0.0152       0      0.0                       0   
951           2  0.0016       0      0.0                       0   
957           4  0.0076       0      0.0                       0   
1034          1  0.0008       0      0.0                       0   
1042          1  0.0006       0      0.0                       0   
...         ...     ...     ...      ...                     ...   
15259         1  0.0008       0      0.0                       0   
15265         8  0.0064       0      0.0                       0   
15323        13  0.0104       0      0.0                       0   
15391         1  0.0008       0      0.0                       0   
15402         6  0.0048       0      0.0                       0   

       post_click_sales_amount  
893                        0.0  
951                        0.0  
957                        0.0  
1034                       0.0  
1042                       0.0  
...                        ...  
15259                      0.0  
15265                      0.0  
15323                      0.0  
15391                      0.0  
15402                      0.0  

[413 rows x 12 columns]

In [85]:
df['placement'].value_counts()

placement
mno    4501
def    3538
ghi    3484
jkl    2504
abc     968
Name: count, dtype: int64

**We have no data for placements of 413 online advertising performances.**
**We can replace them with the mode.**
**For now we would create a new column and store 'No Data Available'**

In [86]:
df['updated_placement'] = df['placement']

In [87]:
df['updated_placement'].fillna('No Data Available',inplace=True)

In [88]:
df['updated_placement'].isnull().sum()

0

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15408 entries, 0 to 15407
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   month                    15408 non-null  object 
 1   day                      15408 non-null  int64  
 2   campaign_number          15408 non-null  object 
 3   user_engagement          15408 non-null  object 
 4   banner                   15408 non-null  object 
 5   placement                14995 non-null  object 
 6   displays                 15408 non-null  int64  
 7   cost                     15408 non-null  float64
 8   clicks                   15408 non-null  int64  
 9   revenue                  15408 non-null  float64
 10  post_click_conversions   15408 non-null  int64  
 11  post_click_sales_amount  15408 non-null  float64
 12  updated_placement        15408 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 1.5+ MB


In [90]:
df[df.duplicated(keep=False)]

month  day campaign_number user_engagement     banner placement  \
5      April    1          camp 1             Low  160 x 600       ghi   
6      April    1          camp 1             Low  160 x 600       mno   
15196   June   29          camp 1            High  800 x 250       ghi   
15197   June   29          camp 1            High  800 x 250       mno   
15198   June   29          camp 3            High  240 x 400       def   
15403  April    1          camp 1             Low  160 x 600       ghi   
15404  April    1          camp 1             Low  160 x 600       mno   
15405   June   29          camp 1            High  800 x 250       ghi   
15406   June   29          camp 1            High  800 x 250       mno   
15407   June   29          camp 3            High  240 x 400       def   

       displays    cost  clicks  revenue  post_click_conversions  \
5            16  0.0249       0   0.0000                       0   
6          2234  0.4044      10   1.8347                       3   
15196         1  0.0157       0   0.0000                       0   
15197         4  0.0123       0   0.0000                       0   
15198      1209  0.3184       2   0.1115                       3   
15403        16  0.0249       0   0.0000                       0   
15404      2234  0.4044      10   1.8347                       3   
15405         1  0.0157       0   0.0000                       0   
15406         4  0.0123       0   0.0000                       0   
15407      1209  0.3184       2   0.1115                       3   

       post_click_sales_amount updated_placement  
5                       0.0000               ghi  
6                     101.7494               mno  
15196                   0.0000               ghi  
15197                   0.0000               mno  
15198                 110.4224               def  
15403                   0.0000               ghi  
15404                 101.7494               mno  
15405                   0.0000               ghi  
15406                   0.0000               mno  
15407                 110.4224               def

**There are 5 rows with duplication.We need to drop these 5 rows.**

In [91]:
df.drop_duplicates(inplace=True)

In [92]:
df.shape

(15403, 13)

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   month                    15403 non-null  object 
 1   day                      15403 non-null  int64  
 2   campaign_number          15403 non-null  object 
 3   user_engagement          15403 non-null  object 
 4   banner                   15403 non-null  object 
 5   placement                14990 non-null  object 
 6   displays                 15403 non-null  int64  
 7   cost                     15403 non-null  float64
 8   clicks                   15403 non-null  int64  
 9   revenue                  15403 non-null  float64
 10  post_click_conversions   15403 non-null  int64  
 11  post_click_sales_amount  15403 non-null  float64
 12  updated_placement        15403 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 1.6+ MB


**Automatic Assessment**
- head and tail
- sample
- info
- isnull
- duplicated
- describe

In [94]:
df.head()

month  day campaign_number user_engagement     banner placement  displays  \
0  April    1          camp 1            High  160 x 600       abc         4   
1  April    1          camp 1            High  160 x 600       def     20170   
2  April    1          camp 1            High  160 x 600       ghi     14701   
3  April    1          camp 1            High  160 x 600       mno    171259   
4  April    1          camp 1             Low  160 x 600       def       552   

       cost  clicks   revenue  post_click_conversions  \
0    0.0060       0    0.0000                       0   
1   26.7824     158   28.9717                      23   
2   27.6304     158   28.9771                      78   
3  216.8750    1796  329.4518                     617   
4    0.0670       1    0.1834                       0   

   post_click_sales_amount updated_placement  
0                   0.0000               abc  
1                1972.4602               def  
2                2497.2636               ghi  
3               24625.3234               mno  
4                   0.0000               def

In [95]:
df.tail()

month  day campaign_number user_engagement    banner placement  \
15398  June   30          camp 3          Medium  728 x 90       def   
15399  June   30          camp 3          Medium  728 x 90       ghi   
15400  June   30          camp 3          Medium  728 x 90       jkl   
15401  June   30          camp 3          Medium  728 x 90       mno   
15402  June   30          camp 3          Medium  728 x 90       NaN   

       displays    cost  clicks  revenue  post_click_conversions  \
15398      5963  0.7884      40   2.2513                       3   
15399       710  0.9867      22   1.2386                       0   
15400       490  0.1686       8   0.4504                       0   
15401     37790  2.6023     195  10.9785                       0   
15402         6  0.0048       0   0.0000                       0   

       post_click_sales_amount  updated_placement  
15398                  69.4256                def  
15399                   0.0000                ghi  
15400                   0.0000                jkl  
15401                   0.0000                mno  
15402                   0.0000  No Data Available

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   month                    15403 non-null  object 
 1   day                      15403 non-null  int64  
 2   campaign_number          15403 non-null  object 
 3   user_engagement          15403 non-null  object 
 4   banner                   15403 non-null  object 
 5   placement                14990 non-null  object 
 6   displays                 15403 non-null  int64  
 7   cost                     15403 non-null  float64
 8   clicks                   15403 non-null  int64  
 9   revenue                  15403 non-null  float64
 10  post_click_conversions   15403 non-null  int64  
 11  post_click_sales_amount  15403 non-null  float64
 12  updated_placement        15403 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 1.6+ MB


**There are a few null values in placement column. But we will be using updated placement column for our analysis.**

In [97]:
df.isnull() 

month    day  campaign_number  user_engagement  banner  placement  \
0      False  False            False            False   False      False   
1      False  False            False            False   False      False   
2      False  False            False            False   False      False   
3      False  False            False            False   False      False   
4      False  False            False            False   False      False   
...      ...    ...              ...              ...     ...        ...   
15398  False  False            False            False   False      False   
15399  False  False            False            False   False      False   
15400  False  False            False            False   False      False   
15401  False  False            False            False   False      False   
15402  False  False            False            False   False       True   

       displays   cost  clicks  revenue  post_click_conversions  \
0         False  False   False    False                   False   
1         False  False   False    False                   False   
2         False  False   False    False                   False   
3         False  False   False    False                   False   
4         False  False   False    False                   False   
...         ...    ...     ...      ...                     ...   
15398     False  False   False    False                   False   
15399     False  False   False    False                   False   
15400     False  False   False    False                   False   
15401     False  False   False    False                   False   
15402     False  False   False    False                   False   

       post_click_sales_amount  updated_placement  
0                        False              False  
1                        False              False  
2                        False              False  
3                        False              False  
4                        False              False  
...                        ...                ...  
15398                    False              False  
15399                    False              False  
15400                    False              False  
15401                    False              False  
15402                    False              False  

[15403 rows x 13 columns]

In [98]:
df[df.duplicated()].shape

(0, 13)

In [99]:
df.describe()

day       displays          cost        clicks       revenue  \
count  15403.000000   15403.000000  15403.000000  15403.000000  15403.000000   
mean      15.518146   15517.383691     11.373902    161.840226     17.935637   
std        8.738737   44398.791785     45.376413    728.389447     96.797025   
min        1.000000       0.000000      0.000000      0.000000      0.000000   
25%        8.000000      78.000000      0.024050      0.000000      0.000000   
50%       15.000000    1182.000000      0.340200      6.000000      0.484200   
75%       23.000000    8969.500000      2.537150     53.000000      3.841800   
max       31.000000  455986.000000    556.704800  14566.000000   2096.211600   

       post_click_conversions  post_click_sales_amount  
count            15403.000000             15403.000000  
mean                42.313965              2123.963529  
std                213.719056             10524.670688  
min                  0.000000                 0.000000  
25%                  0.000000                 0.000000  
50%                  0.000000                 0.000000  
75%                  3.000000               163.502650  
max               3369.000000            199930.318000

**We can change dtypes of some columns to reduce space occupied by the dataframe.**

In [100]:
df['month'].value_counts()

month
April    6267
May      4705
June     4431
Name: count, dtype: int64

In [101]:
df['campaign_number'].value_counts()

campaign_number
camp 3    6916
camp 1    6873
camp 2    1614
Name: count, dtype: int64

In [102]:
df['user_engagement'].value_counts()

user_engagement
Medium    5489
Low       5033
High      4881
Name: count, dtype: int64

In [103]:
df['banner'].value_counts()

banner
728 x 90     3142
300 x 250    2820
240 x 400    2424
670 x 90     2352
468 x 60     1899
160 x 600    1803
580 x 400     606
800 x 250     357
Name: count, dtype: int64

In [104]:
df['updated_placement'].value_counts()

updated_placement
mno                  4499
def                  3537
ghi                  3482
jkl                  2504
abc                   968
No Data Available     413
Name: count, dtype: int64

In [105]:
df['day'].value_counts()

day
8     526
10    526
11    525
12    523
7     521
13    519
5     519
4     517
2     516
6     515
14    515
3     514
9     512
15    510
1     509
20    506
26    504
28    504
18    503
23    500
24    500
17    498
21    498
27    498
25    498
16    497
19    497
29    495
22    493
30    492
31    153
Name: count, dtype: int64

In [106]:
df['displays'].max()

455986

In [107]:
df['clicks'].max()

14566

In [108]:
df['post_click_conversions'].max()

3369

In [109]:
df['cost'].max()

556.7048

In [110]:
df['revenue'].max()

2096.2116

In [111]:
df['post_click_sales_amount'].max()

199930.318

In [112]:
df['month'] = df['month'].astype('category')
df['campaign_number'] = df['campaign_number'].astype('category')
df['user_engagement'] = df['user_engagement'].astype('category')
df['placement'] = df['placement'].astype('category')
df['updated_placement'] = df['updated_placement'].astype('category')
df['day'] = df['day'].astype('int16')
df['clicks'] = df['clicks'].astype('int16')
df['post_click_conversions'] = df['post_click_conversions'].astype('int16')
df['displays'] = df['displays'].astype('int32')
df['cost'] = df['cost'].astype('float16')
df['revenue'] = df['revenue'].astype('float16')
df['post_click_sales_amount'] = df['post_click_sales_amount'].astype('float32')

In [113]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   banner                   15403 non-null  object  
 5   placement                14990 non-null  category
 6   displays                 15403 non-null  int32   
 7   cost                     15403 non-null  float16 
 8   clicks                   15403 non-null  int16   
 9   revenue                  15403 non-null  float16 
 10  post_click_conversions   15403 non-null  int16   
 11  post_click_sales_amount  15403 non-null  float32 
 12  updated_placement        15403 non-null  category
dtypes: category(5), float16(2), float32(1), int16(3), int32(1), object

**We have brought down memory usage from 1.5+ MB to 587.4+ KB.**

**We need to break down banner into banner_width and banner_height.**

In [114]:
df['banner'] = df['banner'].str.split(' x ').str[0].str.strip()

In [115]:
df['banner_width'] = df['banner']

In [116]:
df['banner'] = df_copy['banner']

In [117]:
df['banner_height'] = df['banner'].str.split(' x ').str[1].str.strip()

In [118]:
df.head()

month  day campaign_number user_engagement     banner placement  displays  \
0  April    1          camp 1            High  160 x 600       abc         4   
1  April    1          camp 1            High  160 x 600       def     20170   
2  April    1          camp 1            High  160 x 600       ghi     14701   
3  April    1          camp 1            High  160 x 600       mno    171259   
4  April    1          camp 1             Low  160 x 600       def       552   

         cost  clicks     revenue  post_click_conversions  \
0    0.006001       0    0.000000                       0   
1   26.781250     158   28.968750                      23   
2   27.625000     158   28.984375                      78   
3  216.875000    1796  329.500000                     617   
4    0.067017       1    0.183350                       0   

   post_click_sales_amount updated_placement banner_width banner_height  
0                 0.000000               abc          160           600  
1              1972.460205               def          160           600  
2              2497.263672               ghi          160           600  
3             24625.324219               mno          160           600  
4                 0.000000               def          160           600

In [119]:
df['banner_width'].max()

'800'

In [120]:
df['banner_height'].max()

'90'

In [121]:
df['banner_width'] = df['banner_width'].astype('int64')
df['banner_height'] = df['banner_height'].astype('int64')

In [122]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   banner                   15403 non-null  object  
 5   placement                14990 non-null  category
 6   displays                 15403 non-null  int32   
 7   cost                     15403 non-null  float16 
 8   clicks                   15403 non-null  int16   
 9   revenue                  15403 non-null  float16 
 10  post_click_conversions   15403 non-null  int16   
 11  post_click_sales_amount  15403 non-null  float32 
 12  updated_placement        15403 non-null  category
 13  banner_width             15403 non-null  int64   
 14  banner_heig

In [123]:
df['banner_area'] = df['banner_width'] * df['banner_height']

In [124]:
df.head()

month  day campaign_number user_engagement     banner placement  displays  \
0  April    1          camp 1            High  160 x 600       abc         4   
1  April    1          camp 1            High  160 x 600       def     20170   
2  April    1          camp 1            High  160 x 600       ghi     14701   
3  April    1          camp 1            High  160 x 600       mno    171259   
4  April    1          camp 1             Low  160 x 600       def       552   

         cost  clicks     revenue  post_click_conversions  \
0    0.006001       0    0.000000                       0   
1   26.781250     158   28.968750                      23   
2   27.625000     158   28.984375                      78   
3  216.875000    1796  329.500000                     617   
4    0.067017       1    0.183350                       0   

   post_click_sales_amount updated_placement  banner_width  banner_height  \
0                 0.000000               abc           160            600   
1              1972.460205               def           160            600   
2              2497.263672               ghi           160            600   
3             24625.324219               mno           160            600   
4                 0.000000               def           160            600   

   banner_area  
0        96000  
1        96000  
2        96000  
3        96000  
4        96000

In [125]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   banner                   15403 non-null  object  
 5   placement                14990 non-null  category
 6   displays                 15403 non-null  int32   
 7   cost                     15403 non-null  float16 
 8   clicks                   15403 non-null  int16   
 9   revenue                  15403 non-null  float16 
 10  post_click_conversions   15403 non-null  int16   
 11  post_click_sales_amount  15403 non-null  float32 
 12  updated_placement        15403 non-null  category
 13  banner_width             15403 non-null  int64   
 14  banner_heig

In [126]:
final_copy = df.copy()

In [127]:
df.drop(columns=['banner','placement'],inplace=True)

In [128]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   displays                 15403 non-null  int32   
 5   cost                     15403 non-null  float16 
 6   clicks                   15403 non-null  int16   
 7   revenue                  15403 non-null  float16 
 8   post_click_conversions   15403 non-null  int16   
 9   post_click_sales_amount  15403 non-null  float32 
 10  updated_placement        15403 non-null  category
 11  banner_width             15403 non-null  int64   
 12  banner_height            15403 non-null  int64   
 13  banner_area              15403 non-null  int64   
dtypes: category

In [129]:
export_df = df.copy()

In [130]:
df.to_csv('../dataframes/final_df.csv')

In [131]:
df.to_csv('../dataframes/final_copy_with_initials.csv')

In [132]:
df.to_csv('../dataframes/digital_marketing_after_access_cleaning.csv',index=False)

In [133]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   displays                 15403 non-null  int32   
 5   cost                     15403 non-null  float16 
 6   clicks                   15403 non-null  int16   
 7   revenue                  15403 non-null  float16 
 8   post_click_conversions   15403 non-null  int16   
 9   post_click_sales_amount  15403 non-null  float32 
 10  updated_placement        15403 non-null  category
 11  banner_width             15403 non-null  int64   
 12  banner_height            15403 non-null  int64   
 13  banner_area              15403 non-null  int64   
dtypes: category

In [136]:
df.to_csv('../dataframes/marketing_final.csv',index=False)

**Final Dataframe is ready after data accessing and data cleaning phase.Time for Exploratory Data Analysis(EDA).**

In [139]:
df.to_csv('../dataframes/marketing_digital_after_accessing_cleaning_with_specific_dtypes',index=False,dtypes = {
              'month' : 'category',
              'day' : 'int16',
              'campaign_number' : 'category',
              'user_engagement' : 'category',
              'displays' : 'int32',
              'cost' : 'float16',
              'clicks' : 'int16',
              'revenue' : 'float16',
              'post_click_conversions' : 'int16',
              'post_click_sales_amount' : 'float32',
              'updated_placement' : 'category',
              'banner_width' : 'int64',
              'banner_height' : 'int64',
              'banner_area' : 'int64'
          })

TypeError: NDFrame.to_csv() got an unexpected keyword argument 'dtypes'

In [140]:
df = df.astype({
    'month': 'category',
    'day': 'int16',
    'campaign_number': 'category',
    'user_engagement': 'category',
    'displays': 'int32',
    'cost': 'float16',
    'clicks': 'int16',
    'revenue': 'float16',
    'post_click_conversions': 'int16',
    'post_click_sales_amount': 'float32',
    'updated_placement': 'category',
    'banner_width': 'int64',
    'banner_height': 'int64',
    'banner_area': 'int64'
})

In [141]:
df.to_csv('../dataframes/marketing_digital_after_accessing_cleaning_with_specific_dtypes.csv', index=False)

In [144]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15403 entries, 0 to 15402
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   month                    15403 non-null  category
 1   day                      15403 non-null  int16   
 2   campaign_number          15403 non-null  category
 3   user_engagement          15403 non-null  category
 4   displays                 15403 non-null  int32   
 5   cost                     15403 non-null  float16 
 6   clicks                   15403 non-null  int16   
 7   revenue                  15403 non-null  float16 
 8   post_click_conversions   15403 non-null  int16   
 9   post_click_sales_amount  15403 non-null  float32 
 10  updated_placement        15403 non-null  category
 11  banner_width             15403 non-null  int64   
 12  banner_height            15403 non-null  int64   
 13  banner_area              15403 non-null  int64   
dtypes: category